### PR TITLE
Fix usage of additional kwargs in `from_encoder_decoder_pretrained` in encoder-decoder models

### PR DESCRIPTION
**Pull request by jsnfly**
# What does this PR do?

Somewhat continuing https://github.com/huggingface/transformers/pull/15043 but concerning all of the encoder-decoder models:
If, like in this [example](https://github.com/huggingface/transformers/blob/f71fb5c36e739d8224419bb091b4c16531df829f/examples/pytorch/speech-recognition/README.md?plain=1#L218), extra keyword arguments are used with `from_encoder_decoder_pretrained` it raises an error:
```pthon
TypeError: __init__() got an unexpected keyword argument 'add_adapter'
```
This is because the arguments [should be part of the config](https://github.com/huggingface/transformers/blob/f71fb5c36e739d8224419bb091b4c16531df829f/src/transformers/models/auto/auto_factory.py#L136).

This PR tries to fix this issue.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@patrickvonplaten

### CODE REVIEW

File: tests/test_modeling_encoder_decoder.py [hunk 1]
```diff
@@ -152,11 +152,20 @@ def check_encoder_decoder_model_from_pretrained(
         decoder_input_ids,
         decoder_attention_mask,
         return_dict,
+        from_model_paths=False,
         **kwargs
     ):
         encoder_model, decoder_model = self.get_encoder_decoder_model(config, decoder_config)
-        kwargs = {"encoder_model": encoder_model, "decoder_model": decoder_model, "return_dict": return_dict}
-        enc_dec_model = EncoderDecoderModel.from_encoder_decoder_pretrained(**kwargs)
+        if from_model_paths:
```
Comment (patrickvonplaten) [inline tests/test_modeling_encoder_decoder.py:159]: Could we maybe create a new test function here to make it more readable?

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py [hunk 2]
```diff
@@ -372,6 +372,8 @@ def from_encoder_decoder_pretrained(
 
             if "config" not in kwargs_encoder:
                 encoder_config = AutoConfig.from_pretrained(encoder_pretrained_model_name_or_path)
```
Comment (patrickvonplaten) [inline src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:374]:
```suggestion
                encoder_config, kwargs_encoder = AutoConfig.from_pretrained(encoder_pretrained_model_name_or_path, **kwargs_encoder, return_unused_kwargs=True)
```

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py [hunk 3]
```diff
@@ -372,6 +372,8 @@ def from_encoder_decoder_pretrained(
...
+                encoder_config.update(kwargs_encoder)
```
Comment (patrickvonplaten) [inline src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:375]:
```suggestion
```

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py [hunk 4]
```diff
@@ -382,7 +384,9 @@ def from_encoder_decoder_pretrained(
 
                 kwargs_encoder["config"] = encoder_config
 
-            encoder = AutoModel.from_pretrained(encoder_pretrained_model_name_or_path, *model_args, **kwargs_encoder)
+            encoder = AutoModel.from_pretrained(
+                encoder_pretrained_model_name_or_path, *model_args, config=kwargs_encoder["config"]
```
Comment (patrickvonplaten) [inline src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:388]:
```suggestion
                encoder_pretrained_model_name_or_path, *model_args, config=kwargs_encoder["config"], **kwargs_encoder
```

File: src/transformers/models/encoder_decoder/modeling_encoder_decoder.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/encoder_decoder/modeling_encoder_decoder.py:374]: I think this is actually the cleanest way to fix the problem. This way everything of `kwargs_encoder` that is relevant for the config is removed and everything that is relevant for the model is kept, *e.g.* `local_files_only=True`. This way we can still do:

```python
AutoModel.from_pretrained(..., **kwargs_encoder)
```

later.

**Comment by patrickvonplaten** [comment]
Hey @jsnfly,

sorry for answering so late. Great job on identifying the problem! I left some improvements as comments in the PR. The problem is that if we remove `**kwargs_encoder` from `AutoModel.from_pretrained(...)` we loose functionality - *e.g.* if someone would pass `local_files_only=True`. 

Could you maybe adapt all examples as shown for `EncoderDecoderModel` above? :-) Think then we can merge this one!

**Comment by jsnfly** [comment]
Hey @patrickvonplaten,

thank you for the suggestions, they do indeed make the code a lot cleaner and more functional! Let me know if there are additional improvements.

**Comment by patrickvonplaten** [comment]
> indeed

Great job @jsnfly! Checked it locally and everything works perfectly fine :-)

**Review by patrickvonplaten — APPROVED** [review]

**Pull request merged**